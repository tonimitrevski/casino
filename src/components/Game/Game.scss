@import "../../style/mixins";
@import "../../style/variables";

.Game {
  flex-basis: 50%;
  padding: 10px;
  &__container {
    cursor: pointer;
    padding: 8px;
    position: relative;
    &__ribbon {
      pointer-events: none;
      position: absolute;
      right: 5px;
      top: 5px;
      z-index: 1;
      overflow: hidden;
      width: 75px;
      height: 75px;
      text-align: right;

      &__label {
        font-size: 10px;
        font-weight: 700;
        color: $color-ribon-label-buttons;
        text-transform: uppercase;
        text-align: center;
        line-height: 20px;
        -webkit-transform: rotate(45deg);
        transform: rotate(45deg);
        width: 100px;
        display: block;
        background: $background-ribon-label-buttons;
        box-shadow: 0 3px 10px -5px #000;
        position: absolute;
        top: 19px;
        right: -21px;
      }
    }
    &__jackpot {
      background-color: rgba(0, 0, 0, 0.3);
      position: absolute;
      width: 93%;
      display: flex;
      justify-content: center;
      top: 15px;
      color: $color-ribon-label-buttons;
      padding: 3px 0;
      span {
        font-weight: bold;
      }
    }
  }
  img {
    width: 100%;
  }
  @include medium {
    flex-basis: 33%;
  }
  @include large {
    flex-basis: 25%;
  }
  @include ex-large {
    flex-basis: 20%;
  }
}
