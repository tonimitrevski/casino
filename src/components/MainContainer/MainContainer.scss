.Main-Container {
  padding-top: 70px;
  display: flex;
  flex-wrap: wrap;
  margin: 0 20px;
  visibility: visible;
  opacity: 1;
  &.hidden{
    visibility: hidden;
    opacity: 0;
  }
}
