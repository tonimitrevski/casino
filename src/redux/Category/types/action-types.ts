import {SelectCategoryActionType} from "./select-category-action-type";

export type CategoryActionTypes = SelectCategoryActionType;
