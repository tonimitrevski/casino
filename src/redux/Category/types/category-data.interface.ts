export interface CategoryDataInterface {
  key: string,
  value: string
}
