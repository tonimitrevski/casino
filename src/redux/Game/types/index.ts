import {FetchGamesAction} from "./fetch-games-action.type";

export type GamesActionTypes = FetchGamesAction;
