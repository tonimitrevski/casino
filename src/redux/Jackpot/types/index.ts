import {FetchJackPotAction} from "./fetch-jackpot-action.type";

export type JackpotActionTypes = FetchJackPotAction;
