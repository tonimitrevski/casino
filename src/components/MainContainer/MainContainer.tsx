import React, {useEffect} from "react";
import "./MainContainer.scss"
import Game from "../Game/Game";
import Spinner from "../Spinner/Spinner";
import {connect} from "react-redux";
import {bindActionCreators, Dispatch} from "redux";
import fetchGamesAndJackPots from "./services/fetch-games-and-jack-pots";
import {getGamesSelector} from "../../redux/Game/selectors";
import {StoreType} from "../../redux/store.type";
import {GameStateInterface} from "../../redux/Game/types/game-state-interface";

const MainContainer = (props: {fetchGames: Function , gameStore: GameStateInterface}) => {
  const {fetchGames, gameStore} = props;

  useEffect(() => {
    fetchGames();
  }, [fetchGames]);

  return (
    <div className="Main-Container">
      <Spinner/>
      {renderGames(gameStore)}
    </div>
  );
}

function renderGames(game: GameStateInterface): JSX.Element[] {
  const gamesCount = game.data.length;
  let renderGamesArray = [];
  for (let i = 0; i < gamesCount; i++) {
    renderGamesArray.push(<Game game={game.data[i]} key={i} />);
  }

  return renderGamesArray;
}

const mapStateToProps = (state: StoreType) => ({
  gameStore: getGamesSelector(state)
})

const mapDispatchToProps = (dispatch: Dispatch) => bindActionCreators({
  fetchGames: fetchGamesAndJackPots,
}, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MainContainer);
