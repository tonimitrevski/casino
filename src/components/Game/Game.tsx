import React, {useEffect, useState} from "react";
import "./Game.scss";
import {GameInterface} from "../../core/models/game-interface";
import {connect} from "react-redux";
import {StoreType} from "../../redux/store.type";
import {getJackpotsSelector} from "../../redux/Jackpot/selectors";
import {JackpotStateInterface} from "../../redux/Jackpot/types/jackpot-state-interface";
import {JackpotsAggregate} from "../../core/aggregate/jackpots.aggregate";
import Ribbon from "./Ribbon/Ribbon";
import {Jackpot} from "./Jackpot/Jackpot";
import {FetchImageWorker} from "../../core/services/fetch-image.worker";
import LazyLoad from "react-lazyload";
import Placeholder from "../Placeholder/Placeholder";

export const Game = (props: { game: GameInterface, jackpots: JackpotStateInterface }) => {
  const { game, jackpots } = props;
  const [image, setImage] = useState("");
  useEffect(() => {
    (new FetchImageWorker()).handle(game.image).then(url => {
      setImage(url)
    }).catch(e => console.log(e));
  }, [game.image]);
  return (
    <>
      {image.length > 0 &&
      <div className="Game">
        <LazyLoad once={true} offset={100} height={150} placeholder={<Placeholder/>}>
          <div className="Game__container">
            {showNewRibbon(game)}
            <img data-testid="image" src={image} alt={game.name} onLoad={() => onLoad(image)}/>
            {showJackpot(jackpots.data, game)}
          </div>
          <div className="Game__play-button"></div>
        </LazyLoad>
      </div>
      }
    </>
  );
}

function onLoad(imageUrl: string) {
  URL.revokeObjectURL(imageUrl)
}

function showNewRibbon(game: GameInterface): string | JSX.Element {
  if(!game.categories.includes("new")) {
    return '';
  }
  return <Ribbon type="new" />;
}

function showJackpot(jackpotAggregate: JackpotsAggregate, game: GameInterface) {
  if(!jackpotAggregate.hasOwnProperty(game.id)) {
    return '';
  }
  return (
    <Jackpot amount={jackpotAggregate[game.id].amount} />
  );
}

const mapStateToProps = (state: StoreType) => ({
  jackpots: getJackpotsSelector(state)
})

export default connect(mapStateToProps)(Game);
