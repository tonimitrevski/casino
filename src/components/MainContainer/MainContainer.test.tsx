import React from 'react';
import {mount, shallow} from 'enzyme';
import {GameStateInterface} from "../../redux/Game/types/game-state-interface";
import Game from "../Game/Game";
import {MainContainer} from "./MainContainer";
jest.mock("./../../redux/Services/fetch-games-and-jack-pots/getDataFromWorker.ts")

describe("MainContainer Component", () => {
  test('should show spinner if data is empty', () => {
    const fetchGames = () => {};
    const gameStore: GameStateInterface = {
      data: [],
      pending: true,
      error: null,
      cache: false
    };
    const wrapper = mount(<MainContainer fetchGames={fetchGames} gameStore={gameStore}/>);
    expect(wrapper.find(".loading").length).toBe(1);
    expect(wrapper.find(Game).length).toBe(0);
  });

  test('Should not show spinner if data is not empty', () => {
    const fetchGames = () => {};
    const gameStore: GameStateInterface = {
      data: [{
        "categories": [
          "slots"
        ],
        "name": "Super Safari",
        "image": "https://casino-api-example.s3-eu-west-1.amazonaws.com/images/games/NYXSUPERSAFARI.jpg",
        "id": "NYXSUPERSAFARI"
      },
        {
          "categories": [
            "slots"
          ],
          "name": "Green Lantern",
          "image": "https://casino-api-example.s3-eu-west-1.amazonaws.com/images/games/NYXGREENLANTERN.jpg",
          "id": "NYXGREENLANTERN"
        }],
      pending: false,
      error: null,
      cache: false
    };
    const wrapper = shallow(<MainContainer fetchGames={fetchGames} gameStore={gameStore}/>);
    expect(wrapper.find(Game).length).toBe(2);
  });
});



