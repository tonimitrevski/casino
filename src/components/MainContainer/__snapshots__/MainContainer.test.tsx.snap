// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainContainer Component Should not show spinner if data is not empty 1`] = `
<DocumentFragment>
  <div
    class="loading"
  />
  <div
    class="Main-Container "
  >
    <div
      class="Game"
    >
      Hello World
    </div>
    <div
      class="Game"
    >
      Hello World
    </div>
  </div>
</DocumentFragment>
`;
